lib/puppet/pops/types/p_binary_type.rb

```diff
@@ -94,6 +94,15 @@ def initialize(bin)
       @binary_buffer = (bin.encoding.name == "ASCII-8BIT" ? bin : bin.b).freeze
     end
 
+    #
+    # This method is called from the Psych Yaml serializer
+    # The serializer will call this method to create a value that will be serialized to YAML.
+    # For this specific class we can save the base64 encoded value
+    #
+    def encode_with(psych_encoder)
+      psych_encoder.represent_scalar(nil,to_s)
```

Comment:
Here's my understanding of the persistence store...

We record whether a property change resulted in an "success" event status (where "success" means the property was already insync or puppet made it so), then we pass the `should` (aka desired) value to `Persistence#set_system_value`.

https://github.com/puppetlabs/puppet/blob/5a86a5f79563abbada2b29e01bfed3b05e69db6e/lib/puppet/transaction/resource_harness.rb#L293

If the status wasn't "success", then we record the value that was stored previously or fallback to the event's `previous_value`, as reported by the provider:

https://github.com/puppetlabs/puppet/blob/5a86a5f79563abbada2b29e01bfed3b05e69db6e/lib/puppet/transaction/resource_harness.rb#L286

That value is stored as the "system" value at the end of the run:

https://github.com/puppetlabs/puppet/blob/5a86a5f79563abbada2b29e01bfed3b05e69db6e/lib/puppet/transaction/resource_harness.rb#L112-L113

During the next agent run, we retrieve the current value from the provider, which is recorded as the `previous_value` in the event, and we retrieve the `old_system_value` from `Persistence#get_system_value`:

https://github.com/puppetlabs/puppet/blob/5a86a5f79563abbada2b29e01bfed3b05e69db6e/lib/puppet/transaction/resource_harness.rb#L165

We then compare the values. If they're different, then it means the system drifted/changed between runs, so the event is marked as a corrective change:

https://github.com/puppetlabs/puppet/blob/5a86a5f79563abbada2b29e01bfed3b05e69db6e/lib/puppet/transaction/event.rb#L147

Assuming that's all correct, then with this PR, I think we'll always mark binary property changes as "corrective", because we base64 encode during serialization, but we don't decode on the way back in. So I think, `old_system_value` will never match `previous_value`. I assume we'd need to implement `init_with`, something like:

```ruby
def init_with(psych_coder)
  initialize(self.class.from_base_64(psych_coder.scalar))
end
```

That said, I'm a little hesitant to go this route, because Puppet YAML serialization has been a source of many bugs over the years due to inconsistent `encode_with`/`init_with` implementations, people forgetting to modify `to_yaml_properties` when adding new fields, and accidentally trying to serialize all instance variables. For example https://tickets.puppetlabs.com/browse/PUP-7381

We created `Puppet::Util::PsychSupport` to address some of the inconsistencies, but there's still a bug with catalog deserialization due to class inheritance: https://tickets.puppetlabs.com/browse/PUP-9154

Since we "just" need to determine if a change is corrective or not, could we use the ToStringifiedConverter (or some other converter) to perform a lossy conversion prior to storing it:

```ruby
value = new_system_value(param,
                         param_to_event[pname.to_s],
                         @persistence.get_system_value(resource.ref, pname.to_s)))
@persistence.set_system_value(convert(value))
```

And then do the same conversion on the previous event value when we compare them?

```ruby
# If the values aren't insync, we have confirmed a corrective_change
insync = @property_instance.insync_values?(old_system_value, convert(previous_value))
```

And now that I think about it, I don't think that will work, because we may have changed the representation in a way that the custom type won't understand. We really do need to preserve the data from the previous run and load it during the next run, but without lossy conversions...